plugins/s3_auth/aws_auth_v4.cc

```diff
@@ -176,39 +176,27 @@ canonicalEncode(const String &in, bool isObjectName)
  * @param inLen input character count
  * @return pointer to the trimmed string.
  */
-std::string
+String
 trimWhiteSpacesAndSqueezeInnerSpaces(const char *in, size_t inLen)
 {
   if (nullptr == in || inLen == 0) {
-    return std::string(in, inLen);
-  }
-
-  const char *first = in;
-  while (size_t(first - in) < inLen && isspace(*first)) {
-    first++;
+    return "";
   }
 
-  const char *last = in + inLen - 1;
-  while (last > in && isspace(*last)) {
-    last--;
-  }
+  String in_str = trimWhiteSpaces(String(in, inLen));
+  String out_str;
```

Comment:
You want a `reserve` here to have only one allocation for `out_str`. Otherwise it's possible to have a allocate / deallocate on every `+=`. You'll need to `resize` after to get the string the actual length.